### PR TITLE
prevent player hold invalid AudioCache* pointer

### PR DESCRIPTION
**Pull request by drelaptop**
fix https://github.com/cocos2d/cocos2d-x/issues/18699

> relate https://github.com/cocos2d/cocos2d-x/issues/17619, too

### CODE REVIEW

**Review by PatriceJiang — APPROVED** [review]

File: cocos/audio/apple/AudioEngine-inl.mm [hunk 1]
```diff
@@ -677,6 +677,7 @@ AL_API ALvoid AL_APIENTRY alGenBuffers(ALsizei n, ALuint *bids)
             if (player->_finishCallbak) {
                 auto& audioInfo = AudioEngine::_audioIDInfoMap[audioID];
                 filePath = *audioInfo.filePath;
+                player->setCache(nullptr); // it's safe for player didn't free audio cache
```
Comment (dumganhar) [inline cocos/audio/apple/AudioEngine-inl.mm:680]: 这里设置为 nullptr ，会导致 _audioCache 在 AudioPlayer::rotateBufferThread 中有线程安全访问风险

File: cocos/audio/apple/AudioEngine-inl.mm [hunk 2]
```diff
@@ -710,6 +711,11 @@ AL_API ALvoid AL_APIENTRY alGenBuffers(ALsizei n, ALuint *bids)
 void AudioEngineImpl::uncacheAll()
 {
     _audioCaches.clear();
+    for(auto&& player : _audioPlayers)
+    {
+        // prevent player hold invalid AudioCache* pointer, since all audio caches purged
+        player.second->setCache(nullptr);
```
Comment (dumganhar) [inline cocos/audio/apple/AudioEngine-inl.mm:717]: AudioEngine::uncacheAll/uncache will invoke AudioEngineImpl::stopAll/stop first, it means that AudioPlayer will be destroyed  before AudioEngineImpl::uncacheAll is invoked.

File: cocos/audio/apple/AudioEngine-inl.mm (hunk 1 above)
Comment (drelaptop) [reply cocos/audio/apple/AudioEngine-inl.mm:680]: 此时 player 对应的 rotateBufferThread 应已停止了。执行这里的代码有前置条件：AL 状态结束，并且方法中马上会执行 `delete player`。

File: cocos/audio/apple/AudioEngine-inl.mm (hunk 2 above)
Comment (drelaptop) [reply cocos/audio/apple/AudioEngine-inl.mm:717]: it just means `player->destroy()` called before uncacheAll is invoked, so `player->_audioCache` is wild pointer before player destructed invoked.

Comment (drelaptop) [reply cocos/audio/apple/AudioEngine-inl.mm:717]: at that short time, the `rotateBufferThread ` should been stopped, if not, the invalid cache pointer will case thread crash too.

**Pull request merged**